File: YunChat-Android/app/src/main/kotlin/com/android/yunchat/service/impl/MessageServiceImpl.kt
package com.android.yunchat.service.impl

import com.android.yunchat.core.SocketConstants
import com.android.yunchat.model.MessageInfoModel
import com.android.yunchat.screen.chat.view.ChatViewModel
import com.android.yunchat.service.MessageService
import com.android.yunchat.service.instance.clientServiceInstance
import com.xuexiang.xutil.XUtil
import org.json.JSONObject
import top.chengdongqing.weui.core.utils.showToast

/**
 * @name 消息服务实现类
 * @author yichen9247
 */
class MessageServiceImpl: MessageService {
    /**
     * @name 发送消息
     * @param content 消息内容
     */
    private fun sendChatMessage(
        content: String,
        type: String = "text",
        autoPush: Boolean = false,
        chatViewModel: ChatViewModel,
        event: String = SocketConstants.Events.SEND_MESSAGE
    ) {
        if (content.trim().isEmpty()) {
            return XUtil.getContext().showToast("内容不能为空")
        }
        clientServiceInstance.sendSocketEmit<MessageInfoModel>(event, chatViewModel, JSONObject().apply {
            put("obj", 0)
            put("type", type)
            put("content", content)
            put("tar", chatViewModel.currentGroupId.intValue)
        }) { response ->
            if (!autoPush) return@sendSocketEmit
            response?.data?.let {
                chatViewModel.messageList.add(it)
            }
        }
        chatViewModel.inputValue.value = ""
    }

    /**
     * @name 发送文本消息
     * @param content 消息内容
     */
    override fun sendTextMessage(content: String, chatViewModel: ChatViewModel) {
        sendChatMessage(
            autoPush = false,
            content = content,
            chatViewModel = chatViewModel
        )
    }

    /**
     * @name 发送文件消息
     * @param content 消息内容
     */
    override fun sendFileMessage(content: String, chatViewModel: ChatViewModel) {
        sendChatMessage(
            type = "file",
            autoPush = false,
            content = content,
            chatViewModel = chatViewModel
        )
    }

    /**
     * @name 发送图片消息
     * @param content 消息内容
     */
    override fun sendImageMessage(content: String, chatViewModel: ChatViewModel) {
        sendChatMessage(
            type = "image",
            autoPush = false,
            content = content,
            chatViewModel = chatViewModel
        )
    }

    /**
     * @name 发送文本消息
     * @param content 消息内容
     */
    override fun sendVideoMessage(content: String, chatViewModel: ChatViewModel) {
        sendChatMessage(
            type = "video",
            autoPush = false,
            content = content,
            chatViewModel = chatViewModel
        )
    }

    /**
     * @name 发送智能消息
     * @param content 消息内容
     */
    override fun sendAgentMessage(content: String, chatViewModel: ChatViewModel) {
        sendChatMessage(
            autoPush = true,
            content = content,
            chatViewModel = chatViewModel,
            event = SocketConstants.Events.SEND_AI_MESSAGE
        )
    }

    override fun getMessageType(
        length: Int,
        type: String,
        content: String
    ): String {
        return when (type) {
            "text" -> content.take(length)
            "file" -> "[文件消息]"
            "image" -> "[图片消息]"
            "video" -> "[视频消息]"
            "invite" -> "[分享群聊]"
            else -> "[未知消息]"
        }
    }
}

File: YunChat-Client/src/service/impl/MessageServiceImpl.ts
import utils from "@/scripts/utils";
import { EventType } from "@/enums/EventType";
import { ToastType } from "@/enums/ToastType";
import { MessageService } from "../MessageService";
import { SocketIOClient } from "@/socket/SocketIOClient";
import { groupInfoType, messageType, restfulType } from "../../../types";
import { socketIOClientInstance } from "../ServiceInstance";

/**
 * @name 消息服务实现类
 * @author yichen9247
 */
export class MessageServiceImpl implements MessageService {
    // 统一发送消息目标
    private sendMessage(
        event: string,
        content: string,
        type: 'text' | 'image' | 'file' | 'video' | 'invite',
        callback?: (response: restfulType<any>) => void,
    ): void {
        if (!this.validateContent(content)) return;
        socketIOClientInstance.sendSocketEmit({
            event, data: { type, content, 
                obj: 0,
                tar: SocketIOClient.applicationStore.groupInfo.gid
            },
            callback: this.createResponseHandler(callback)
        });
    }

    // 统一消息内容校验
    private validateContent(content: string): boolean {
        const trimmedContent = content?.trim();

        if (!trimmedContent) {
            utils.showToasts(ToastType.WARNING, '内容不能为空');
            return false;
        }
        return true;
    }

    // 统一消息响应处理
    private createResponseHandler(callback?: (response: restfulType<messageType>) => void) {
        return async (response: restfulType<any>): Promise<void> => {
            if (response.code !== 200) {
                utils.showToasts(ToastType.ERROR, response.message);
                return;
            }
            callback?.(response);
        };
    }

    // 获取用户消息类型
    private readonly MESSAGE_TYPE_MAP: Map<string, string> = new Map([
		['text', '文本消息'],
		['file', '文件消息'],
		['image', '图片消息'],
		['video', '视频消息'],
        ['invite', '群聊邀请'],
	]);

    /**
     * @name 获取消息类型
     * @param type 消息类型
     */
    getMessageType(type: string): string {
        return this.MESSAGE_TYPE_MAP.get(type) || '未知消息';
    }

    /**
     * @name 获取消息类型
     * @param type 消息类型
     */
    getGroupMessageType(type: string, content: string): string {
        if (type === "text") return content.slice(0, 8);
        return `[${this.MESSAGE_TYPE_MAP.get(type) || '未知消息'}]`
    }

    /**
     * @name 发送文本消息
     * @param content 内容
     * @param callback 回调
     */
    sendTextMessage(content: string, callback?: (response: restfulType<messageType>) => void): void {
        this.sendMessage(EventType.SEND_MESSAGE, content, 'text', callback);
    }

    /**
     * @name 发送图片消息
     * @param content 内容
     * @param callback 回调
     */
    sendImageMessage(content: string, callback?: (response: restfulType<messageType>) => void): void {
        this.sendMessage(EventType.SEND_MESSAGE, content, 'image', callback);
    }

    /**
     * @name 发送文件消息
     * @param content 内容
     * @param callback 回调
     */
    sendFileMessage(content: string, callback?: (response: restfulType<messageType>) => void): void {
        this.sendMessage(EventType.SEND_MESSAGE, content, 'file', callback);
    }

    /**
     * @name 发送视频消息
     * @param content 内容
     * @param callback 回调
     */
    sendVideoMessage(content: string, callback?: (response: restfulType<messageType>) => void): void {
        this.sendMessage(EventType.SEND_MESSAGE, content, 'video', callback);
    }

    /**
     * @name 发送智能消息
     * @param content 内容
     * @param callback 回调
     */
    sendAgentMessage(content: string, callback?: (response: restfulType<messageType>) => void): void {
        this.sendMessage(EventType.SEND_AI_MESSAGE, content, 'text', callback);
    }

    /**
     * @name 发送群聊邀请
     * @param gid 群聊id
     * @param callback 回调
     */
    sendGroupInvitedMessage(group: groupInfoType, callback?: (response: restfulType<messageType>) => void): void {
        group.notice = null;
        group.member.length = 0;
        this.sendMessage(EventType.SEND_MESSAGE, JSON.stringify(group), 'invite', callback);
    }

    getNewMessageType( length: number, type: string, content: string): string {
        switch (type) {
            case "text":
                return content.substring(0, length);
            case "file":
                return "发送了文件消息";
            case "image":
                return "发送了图片消息";
            case "video":
                return "发送了视频消息";
            case "invite":
                return "发送了分享群聊";
            default:
                return "发送了未知消息";
        }
    }
}